[Playlist] UI Updates

duduBTW opened the pull request:
## UI UPDATES

---

https://github.com/user-attachments/assets/fd5a3aec-31ee-42cc-a72d-467456af77ec



- Add support for images on the playlists.
- Change `playlist::resetDropdown` to  `query::playlistDropdown`, that way we can return more info than just names.
- Add `playlist::resetDropdown` listner.
- Add `query::playlistDropdown` router.
- Add direction for roving group, horizontal and vertial.
- Add dropdown sub.
- Update playlist design.

duduBTW committed:
- Add support for images on the playlists.
- Change `playlist::resetDropdown` to  `query::playlistDropdown`, that way we can return more info than just names.
- Add `playlist::resetDropdown` listner.
- Add `query::playlistDropdown` router.
- Add direction for roving group, horizontal and vertial.
- Add dropdown sub.
- Update playlist design.

duduBTW committed:
Merge branch 'master' of https://github.com/D0m1nos/osu-radio

D0m1nos commented on src/main/router/playlist-router.ts:
```diff
@@ -146,30 +155,29 @@ Router.respond("playlist::remove", async (_evt, playlistName, song) => {
-    return fail("Playlist already exists");
+  if (isPlaylistNameTaken(playlistNames, oldName)) {
+    playlists.delete(oldName);
   }
```
If i have two playlists "a" and "b" and i rename "b" to "a" the first playlist ("a") gets deleted completely.
I think it's better to just show an error ("Playlist name already in use") because someone could mistakenly lose a playlist like this.

duduBTW replied:
This is not the "Playlist name already in use" check, it simply is deleting the old playlist when you rename it. But this is a good point, I added the following validation above.

``` ts
if (isPlaylistNameTaken(playlistNames, playlist.name)) {
    return fail("Playlist already exists");
}
```

This should ensure the user don't overwrite an already existing playlist.

On that topic, relying only on the name to make edit or delete playlists is not a good practice, we should have an unique ID for every playlist, maybe something we can add on the future.

D0m1nos replied:
You're right, we'll think about it when we change to a DBMS

D0m1nos commented on src/renderer/src/components/playlist/new-playlist/NewPlaylist.tsx:
When creating a playlist the first designs had a separate page (like this one), then it was changed (discord discussion [here](https://discord.com/channels/1284644086820896879/1289979431145701498/1294771677456105597)) because it has a bit of unused space and on big screens clicking the button all the way at the bottom is a bit inconvenient.
I like having it on a separate page but i think the button needs to be under the playlist name input box.

duduBTW replied:
The solution to removing the white space is what we discussed, having a card instead of a separate page. However, since the design for this card is not complete yet (there's no edit functionality), we will have to stick with the separate page for now. 

The button on the bottom is not that big of a deal because the user will just use the enter to create or edit most of the time, and white space is not something inherently bad on design.

D0m1nos replied:
Got it. Thanks.

D0m1nos commented on src/renderer/src/components/playlist/playlist-title/PlaylistTitle.tsx:
```diff
@@ -0,0 +1,25 @@
+      >
+        <ArrowLeftIcon class="text-subtext" size={20} />
+      </Button>
+      <h3 class="text-xl uppercase">{props.title}</h3>
```
I don't really like the title being in uppercase. Do you think it's ok? (sorry i don't know much about design)

duduBTW replied:
We are using upper case on the title to have hierarchical clarity on subpages, uppercase text can differentiate titles or headers from body text, aiding visual hierarchy.

D0m1nos replied:
Interesting, thanks for the explanation!

D0m1nos commented on src/renderer/src/components/song/context-menu/items/PlaylistChooser.tsx:
```diff
@@ -1,91 +1,77 @@
-      }}
-      class="max-h-72 w-40 overflow-auto pr-1.5"
-    >
+    <div class="flex flex-col gap-1">
```
If you have a lot of playlists (or a small window) this doesn't scroll, initially i had fixed it by adding `h-screen overflow-y-scroll`, but i noticed that at the top (on the title bar) you can't click the items in the popover, so i used `h-72 overflow-y-scroll`.

duduBTW replied:
Added the scroll back.

D0m1nos commented on src/renderer/src/components/song/context-menu/items/PlaylistChooser.tsx:
```diff
@@ -1,91 +1,77 @@
+      <div class="flex items-center gap-2">
+        <SongImage
+          class="size-7 flex-shrink-0 rounded bg-cover"
+          src={props.playlist.image}
```
This only shows user uploaded images, it doesn't show the image of the first song in the playlist.

duduBTW replied:
Fixed.

D0m1nos commented on src/renderer/src/components/playlist/playlist-song-list/PlaylistSongList.tsx:
```diff
@@ -1,76 +1,74 @@
           class="rounded-lg"
           onClick={() => {
-            setEditMode(!editMode());
+            setPlaylistActivePage({
+              name: "edit",
+              playlist: props.playlist,
+              from: { name: "songs", playlist: props.playlist },
+            });
```
If you click the edit button in this page and change the name of the playlist it puts you back in a page with the old playlist name.
If you edit the image of the playlist and then re-click the edit button you will see the old image.

duduBTW replied:
Fixed this issue.

D0m1nos reviewed (changes requested):
Thank you again for this pr! I marked some things that i noticed while testing, the rest looks very good.

D0m1nos commented on src/renderer/src/components/song/context-menu/items/PlaylistChooser.tsx:
```diff
@@ -1,91 +1,77 @@
+    >
+      <div class="flex items-center gap-2">
+        <SongImage
+          class="size-7 flex-shrink-0 rounded bg-cover"
```
This is missing `bg-center`

D0m1nos commented on src/main/router/dir-router.ts:
```diff
@@ -3,6 +3,19 @@ import { none, some } from "../lib/rust-like-utils-backend/Optional";
+    return none();
+  }
+
+  return some(result.filePaths[0]);
```
Is using the path of the image ok? Wouldn't it be better to copy the image in the application storage and use that path instead?

duduBTW replied:
Copying the image would take more space on the user device unnecessarily.

D0m1nos replied:
Right, i was worried that when deleting the image locally it wouldn't load anymore but i guess it's not a big problem

duduBTW replied:
This is the down side of this approach, we lose the reference to the image if it is deleted or moved, but the fact that we don't have to copy the image far outweighs this.

D0m1nos reviewed (changes requested):
Sorry i just noticed a couple of things

Commit:
- Made so the playlist thumb image can come from the song.
- Added scroll to playlist chooser.
- Update song page when a playlist is edited.

Commit:
- Checks if playlist already exist.

D0m1nos commented:
Thanks for resolving those issues. One last thing, when clicking on "create playlist" from the context menu the song doesn't get automatically added anymore, could you re-add it?
